File: Dart/6-LinkedList/Problems/3-LinkedListCycle/LinkedListCyclePro.dart
import 'dart:collection';

class ListNode {
  int val;
  ListNode? next;
  ListNode(this.val, [this.next]);
}

void main() {
  // Creating a linked list with a cycle
  ListNode head = ListNode(3);
  head.next = ListNode(2);
  head.next!.next = ListNode(0);
  head.next!.next!.next = ListNode(-4, head.next); // Cycle here

  print("Detect Cycle (Floyd's Algorithm): ${hasCycleFloyd(head)}");
  print("Detect Cycle (HashSet Storage): ${hasCycleHashSet(head)}");
  print("Detect Cycle (Two-Pointer Optimization): ${hasCycleTwoPointer(head)}");
  print("Detect Cycle (Node Marking): ${hasCycleMarking(head)}");
  print("Detect Cycle (Cycle Length Calculation): ${hasCycleLength(head)}");
  print("Detect Cycle (Reverse Linked List Technique): ${hasCycleReverse(head)}");
}

// Approach 1: Floyd’s Cycle Detection Algorithm (Tortoise and Hare)
// Explanation:
// - Uses two pointers, slow and fast.
// - Slow moves one step at a time, fast moves two steps.
// - If there's a cycle, the fast pointer will eventually catch up to the slow pointer.
// - If there's no cycle, the fast pointer will reach the end of the list (null).
bool hasCycleFloyd(ListNode? head) {
  ListNode? slow = head, fast = head;
  while (fast != null && fast.next != null) {
    slow = slow?.next; // Slow pointer moves one step
    fast = fast.next?.next; // Fast pointer moves two steps
    if (slow == fast) return true; // Cycle detected if pointers meet
  }
  return false; // No cycle found
}

// Approach 2: Using HashSet to Store Visited Nodes
// Explanation:
// - Uses a HashSet to store visited nodes.
// - Iterates through the list, adding each node to the set.
// - If a node is already in the set, it means a cycle is present.
bool hasCycleHashSet(ListNode? head) {
  HashSet<ListNode> visited = HashSet<ListNode>();
  while (head != null) {
    if (visited.contains(head)) return true; // Cycle detected if node is already visited
    visited.add(head); // Add node to the visited set
    head = head.next; // Move to the next node
  }
  return false; // No cycle found
}

// Approach 3: Two-Pointer Optimization (Fast-Slow Pointers)
// Explanation:
// - Similar to Floyd's Algorithm, but initializes fast pointer to head.next.
// - This avoids the initial check of slow == fast when both are at head.
// - Otherwise, the logic is the same: slow moves one step, fast moves two steps.
bool hasCycleTwoPointer(ListNode? head) {
  if (head == null || head.next == null) return false; // No cycle if list is empty or has only one node
  ListNode? slow = head, fast = head.next; // Initialize fast pointer to head.next
  while (fast != null && fast.next != null) {
    if (slow == fast) return true; // Cycle detected if pointers meet
    slow = slow?.next; // Slow pointer moves one step
    fast = fast.next?.next; // Fast pointer moves two steps
  }
  return false; // No cycle found
}

// Approach 4: Marking Nodes with a Special Value (Destructive Method)
// Explanation:
// - Modifies the linked list by marking visited nodes with a special value (999999 in this case).
// - If a node with the special value is encountered, it means a cycle is present.
// - This method modifies the original list, which might not be desirable in all cases.
bool hasCycleMarking(ListNode? head) {
  while (head != null) {
    if (head.val == 999999) return true; // Cycle detected if node is marked
    head.val = 999999; // Mark the node
    head = head.next; // Move to the next node
  }
  return false; // No cycle found
}

// Approach 5: Detect Cycle by Finding Cycle Length
// Explanation:
// - Uses Floyd's Algorithm to detect a cycle.
// - If a cycle is found, calculates the length of the cycle.
// - Starts from the meeting point of slow and fast pointers and counts the nodes until it returns to the meeting point.
bool hasCycleLength(ListNode? head) {
  ListNode? slow = head, fast = head;
  while (fast != null && fast.next != null) {
    slow = slow?.next; // Slow pointer moves one step
    fast = fast.next?.next; // Fast pointer moves two steps
    if (slow == fast) {
      int length = 1; // Initialize cycle length
      ListNode? temp = slow?.next; // Start from the next node of the meeting point
      while (temp != null && temp != slow) {
        length++; // Increment cycle length
        temp = temp.next; // Move to the next node
      }
      print("Cycle Length: $length"); // Print the cycle length
      return true; // Cycle found
    }
  }
  return false; // No cycle found
}

// Approach 6: Reversing the Linked List (If Cycle Exists, Head Remains Unchanged)
// Explanation:
// - Reverses the linked list.
// - If there's a cycle, the head of the original list will be encountered during reversal.
// - If there's no cycle, the reversal will complete without encountering the original head.
// - This method also modifies the linked list.
bool hasCycleReverse(ListNode? head) {
  ListNode? prev = null;
  ListNode? current = head;
  while (current != null) {
    ListNode? nextNode = current.next; // Store the next node
    current.next = prev; // Reverse the pointer
    prev = current; // Move prev to the current node
    current = nextNode; // Move current to the next node
    if (current == head) return true; // Cycle detected if the original head is encountered
  }
  return false; // No cycle found
}

File: Java/6-LinkedList/Problems/3-LinkedListCycle/LinkedListCyclePro.java
import java.util.HashSet;
import java.util.Set;

/**
 * This class provides multiple solutions to detect a cycle in a linked list.
 * Each approach is commented with its real-world application scenario and time/space complexity.
 */
public class LinkedListCycleDetector {

    // Definition for singly-linked list node
    static class ListNode {
        int val;
        ListNode next;
        ListNode(int x) {
            val = x;
            next = null;
        }
    }

    /**
     * Solution 1: Hash Table Approach
     * Real-world use: Ideal for scenarios where you need to track visited nodes for other purposes
     * (like debugging or analytics) while detecting cycles.
     *
     * Time Complexity: O(n) - We visit each node exactly once
     * Space Complexity: O(n) - We store each node in the hash set
     */
    public static boolean hasCycleHashTable(ListNode head) {
        // Create a set to store visited nodes
        Set<ListNode> visitedNodes = new HashSet<>();

        // Traverse the list
        ListNode current = head;
        while (current != null) {
            // If we've seen this node before, there's a cycle
            if (visitedNodes.contains(current)) {
                return true;
            }
            // Add current node to the set
            visitedNodes.add(current);
            // Move to next node
            current = current.next;
        }

        // If we reach null, no cycle found
        return false;
    }

    /**
     * Solution 2: Floyd's Tortoise and Hare (Two Pointers)
     * Real-world use: Most efficient for memory-constrained environments as it uses O(1) space.
     * Commonly used in production systems for cycle detection.
     *
     * Time Complexity: O(n) - Linear time
     * Space Complexity: O(1) - Constant space
     */
    public static boolean hasCycleTwoPointers(ListNode head) {
        if (head == null || head.next == null) {
            return false;
        }

        // Initialize slow and fast pointers
        ListNode slow = head;
        ListNode fast = head.next;

        while (slow != fast) {
            // If fast reaches the end, no cycle
            if (fast == null || fast.next == null) {
                return false;
            }
            // Move slow by 1 and fast by 2
            slow = slow.next;
            fast = fast.next.next;
        }

        // If slow and fast meet, cycle exists
        return true;
    }

    /**
     * Solution 3: Marking Visited Nodes (Destructive)
     * Real-world use: Useful in temporary debugging scenarios where you can modify the list structure
     * and don't need to preserve the original list.
     *
     * Time Complexity: O(n) - Linear time
     * Space Complexity: O(1) - Constant space, but modifies the original list
     */
    public static boolean hasCycleMarking(ListNode head) {
        // Traverse the list
        ListNode current = head;
        while (current != null) {
            // If we find a marked node, cycle exists
            if (current.val == Integer.MIN_VALUE) {
                return true;
            }
            // Mark the current node by changing its value
            current.val = Integer.MIN_VALUE;
            // Move to next node
            current = current.next;
        }

        // No marked nodes found, no cycle
        return false;
    }

    /**
     * Solution 4: Reverse List Approach
     * Real-world use: Useful when you need to both detect cycles and potentially reverse the list
     * if no cycle exists. More of a academic approach than production code.
     *
     * Time Complexity: O(n) - Linear time
     * Space Complexity: O(1) - Constant space
     */
    public static boolean hasCycleReverseList(ListNode head) {
        if (head == null || head.next == null) {
            return false;
        }

        ListNode reversed = null;
        ListNode current = head;
        ListNode next;

        try {
            while (current != null) {
                // Store next node
                next = current.next;
                // Reverse the link
                current.next = reversed;
                // Move pointers
                reversed = current;
                current = next;

                // If we reach head again, cycle exists
                if (current == head) {
                    return true;
                }
            }
        } finally {
            // In a real implementation, you might want to restore the list
            // This is omitted here for simplicity
        }

        // If we successfully reversed the whole list, no cycle
        return false;
    }

    /**
     * Solution 5: Using a Temp Node
     * Real-world use: Useful when you can't modify node values (like in readonly systems)
     * but can temporarily modify pointers. Less destructive than Solution 3.
     *
     * Time Complexity: O(n) - Linear time
     * Space Complexity: O(1) - Constant space
     */
    public static boolean hasCycleTempNode(ListNode head) {
        // Create a temporary node
        ListNode temp = new ListNode(0);

        ListNode current = head;
        while (current != null) {
            // If we find a node pointing to temp, cycle exists
            if (current.next == temp) {
                return true;
            }

            // Store next node
            ListNode next = current.next;
            // Point current node to temp
            current.next = temp;
            // Move to next node
            current = next;
        }

        // No cycles found
        return false;
    }

    // Helper method to create a linked list with cycle for testing
    public static ListNode createLinkedListWithCycle(int[] values, int pos) {
        if (values == null || values.length == 0) {
            return null;
        }

        ListNode[] nodes = new ListNode[values.length];
        ListNode head = new ListNode(values[0]);
        nodes[0] = head;

        for (int i = 1; i < values.length; i++) {
            nodes[i] = new ListNode(values[i]);
            nodes[i-1].next = nodes[i];
        }

        if (pos >= 0 && pos < values.length) {
            nodes[values.length-1].next = nodes[pos];
        }

        return head;
    }

    // Test cases
    public static void main(String[] args) {
        // Test case 1: No cycle
        ListNode list1 = createLinkedListWithCycle(new int[]{1, 2, 3, 4}, -1);
        System.out.println("Test Case 1 (No cycle):");
        System.out.println("Hash Table: " + hasCycleHashTable(list1));
        System.out.println("Two Pointers: " + hasCycleTwoPointers(list1));
        System.out.println("Marking: " + hasCycleMarking(createLinkedListWithCycle(new int[]{1, 2, 3, 4}, -1)));
        System.out.println("Reverse List: " + hasCycleReverseList(createLinkedListWithCycle(new int[]{1, 2, 3, 4}, -1)));
        System.out.println("Temp Node: " + hasCycleTempNode(createLinkedListWithCycle(new int[]{1, 2, 3, 4}, -1)));

        // Test case 2: Has cycle
        ListNode list2 = createLinkedListWithCycle(new int[]{1, 2, 3, 4}, 1);
        System.out.println("\nTest Case 2 (Has cycle):");
        System.out.println("Hash Table: " + hasCycleHashTable(list2));
        System.out.println("Two Pointers: " + hasCycleTwoPointers(list2));
        System.out.println("Marking: " + hasCycleMarking(createLinkedListWithCycle(new int[]{1, 2, 3, 4}, 1)));
        System.out.println("Reverse List: " + hasCycleReverseList(createLinkedListWithCycle(new int[]{1, 2, 3, 4}, 1)));
        System.out.println("Temp Node: " + hasCycleTempNode(createLinkedListWithCycle(new int[]{1, 2, 3, 4}, 1)));
    }
}